torch/autograd/gradcheck.py

```diff
@@ -160,18 +161,41 @@ def get_stride(size):
     return jacobian
 
 
-def get_analytical_jacobian(input, output, nondet_tol=0.0, grad_out=1.0):
-    # it is easier to call to_dense() on the sparse output than
-    # to modify analytical jacobian
-    if output.is_sparse:
-        raise ValueError('Sparse output is not supported at gradcheck yet. '
-                         'Please call to_dense() on the output of fn for gradcheck.')
-    if output.layout == torch._mkldnn:  # type: ignore
-        raise ValueError('MKLDNN output is not supported at gradcheck yet. '
-                         'Please call to_dense() on the output of fn for gradcheck.')
-    diff_input_list = list(iter_tensors(input, True))
-    jacobian = make_jacobian(input, output.numel())
-    jacobian_reentrant = make_jacobian(input, output.numel())
+def check_analytical_jacobian(inputs, output, nondet_tol, grad_out, check_batched_grad,
+                              check_grad_dtypes, raise_exception):
+    jacobian, reentrant, correct_grad_sizes, correct_grad_types = \
+        get_analytical_jacobian(inputs, output, nondet_tol, grad_out)
+
+    complex_str = '(calculated using complex valued grad output) ' \
+        if isinstance(grad_out, complex) else ''
+
+    def fail_test(msg):
+        if raise_exception:
+            raise RuntimeError(msg)
+
+    if not correct_grad_types and check_grad_dtypes:
+        fail_test('Gradient{0} has dtype mismatch'.format(complex_str))
```

Comment:
nit: We can use f-strings to make this more concise. e.g. `fail_test(f'Gradient{complex_str} has dtype mismatch')`. Ditto for the other strings in this file.